Feature/multiple request

serpient opened the pull request:
# Testing the Feature
* To test the simultaneous client sessions with a benchmark, start App.main() and then run `ab -r -n 5000 -c 1000 http://localhost:1234/` in another terminal.

* To test the simultaneous client sessions in a real life session, start App.main() and then open at least 2 terminals and run `nc localhost 1234`. From here, you can type an input, and the session should echo back your input and then close itself.

# What was Done
To achieve this feature, the code base had to be restructured to have a main thread in App that handles the server creation, a background thread (Server Thread) that runs the EchoProtocol to wait for new client connections, which then runs EchoSessions in new threads. An important part to get this working was to make sure EchoSessions closes after each call.

# Benchmark Result of 1500 Simultaneous Requests with Apache AB
<img width="677" alt="Screen Shot 2019-06-06 at 8 49 04 AM" src="https://user-images.githubusercontent.com/29721784/59047037-f5354a80-8837-11e9-94d1-ca5ef7e59160.png">

serpient committed:
Refactor echo server to prepare for threading

- move client creation and input/output stream into Client Socket
- move server stop in echoServer
- Remove Client.java and instead use a ClientSocket, which is a duck type of SocketWrapper interface. There is also MockClientSocket for testing
- Create a server socket interface with 2 duck types -- echoServerSocket and MockServerSocket
- adjust tests for new setup procedures
- currently failing tests for server start/stop as its infinity looping. Will need to research threading

serpient committed:
Add threads to start and stop a client session

- App currently creates a server socket, which then creates a client when another terminal runs `nc localhost 1234`. It starts the echo server functionality, and the thread is stopped with either client input of "Stop" or if it overruns the timer currently set in App. This functionality is a draft version to make sure we can start and stop a session remotely. It will be replaced with a dedicated start and stop method thats not reliant on a timer.
- SocketWrapper and its duck types of ClientSocket and MockClientSocket has a new ready() method that checks to see if the input stream is ready. This is used as a way to control the input stream read() function, which currently blocks the thread while its waiting for user input.
- Rename Echo Server as Echo Session. It implements Runnable and all functionality has been moved to run().
- Update logic in EchoSession.run() so that it can be stopped using Thread.interrupt() from main App.

serpient committed:
Remove gradle folder from git history

- Not good practice to check in the gradle executable and jar file. "This is considered a security issue since someone can alter them to run malicious code. The best practice is to leave it out of your repository, and document gradle as a dependency that users can install onto their own computers."

serpient committed:
Add multiple simultaneous client session feature

* To test the simultaneous client sessions with a benchmark, start App.main() and then run `ab -r -n 5000 -c 1000 http://localhost:1234/` in another terminal.
* To test the simultaneous client sessions in a rea life session, start App.main() and then open at least 2 terminals and run `nc localhost 1234`. From here, you can type an input, and the session should echo back your input and then close itself.

* To achieve this feature, the code base had to be restructured to have a main thread in App that handles the server creation, a background thread (Server Thread) that runs the EchoProtocol to wait for new client connections, which then runs EchoSessions in new threads. An important part to get this working was to make sure EchoSessions closes after each call.

- Update App to handle creation of a server socket and thread, then starting the server thread.
- Create a Runnable EchoProtocol class that waits for new clients to accept, then create and start a new client session thread. It has a timeout backup to close the session if it's taking too long.
- Remove ServerSocketWrapper and the 2 duck types of EchoServerSocket and MockServerSocket. It's not needed anymore to complete the feature.
- Update EchoSession to close the session after client input of "Stop" and after it echo's back a client input. Closing the session after a single echo is critical to testing with Apache AB -- If it does not close, AB will run into a timeout error because it will continue to wait for the client to complete. Closing the session after a call also matches what a real HTTP Server would behave like.
- Update ServerTest to check for run, then close the session after the assertion.

serpient committed:
Remove unused code for multiple request feature

By having the client session close itself after it responds, we no longer need to have explicit external client session stop/interrupt controls. These series of changes removes the code that relates to controlling a continuously running EchoSession, as well as remove any other unused code.

-  Clean up SocketWrapper interface and it's 2 duck types of ClientSocket and MockClientSocket. Remove unused methods of getInput and getOutput, as well as the ready() method.
* Ready() was needed previously to check whether inputStream was ready to be read from. Without this, the inputStream would continuously wait for input and block the thread and blocking any attempts to interrupt/stop the thread. Now that we are closing the ClientSession after a single response, inputStream is no longer infinitely waiting.

- Update EchoProtocol and streamline run().
* Remove the timer that kills client sessions if it runs too long as ClientSession now closes itself after a single response.
* Removed the usage of threadGroup as I was previously using it to try and gauge the number of active threads happening at a single time. While it works for user testing, it would not count the threads accurately during AB apache testing.

- Update EchoSession to simplify run().
* As EchoSession closes after a single echo, we no longer need to check whether client is ready to read from and putting it to sleep every second. This was only used during a continuously running EchoSession, to make sure that Input Stream would not block us from closing the session externally.
* Remove unnecessary client input check of "Stop" to stop the session.

ssunday commented on src/main/java/java_echo_server/App.java:
```diff
@@ -1,57 +1,22 @@
-            serverSocket.close();
-        } catch (IOException e) {
-            System.err.println("The server could not be shut down.");
+        private static int setPortNumber(String[] terminal_args) {
```
more spacing here than elsewhere—why?

serpient replied:
I tabbed the ternary condition so that it would be read under terminal_args and not return.
<img width="512" alt="Screen Shot 2019-06-06 at 10 37 35 AM" src="https://user-images.githubusercontent.com/29721784/59053817-1ea9a280-8847-11e9-852e-a57ebe9b2ec2.png">
<img width="496" alt="Screen Shot 2019-06-06 at 10 37 40 AM" src="https://user-images.githubusercontent.com/29721784/59053823-21a49300-8847-11e9-91d6-0435be5b095c.png">

But this is a very trivial reason and I'll adjust it back.

ssunday replied:
Sorry, I meant the entire file seemed like it had more spacing

serpient replied:
Ahhhh, Ok -- Adjusting!

aaizuss commented:
I appreciate the detail in your commits! In general though, you don't need to be so specific (but I'm definitely not complaining about it because it's certainly helpful when reviewing your code). That level of detail is often more useful in your pull request description. On my current client, my team tends to put technical explanations/summaries/reasoning behind certain decisions in PR descriptions.

aaizuss commented on src/main/java/java_echo_server/App.java:
```diff
@@ -1,57 +1,22 @@
-        } catch (IOException e) {
-            System.err.println(e);
-            return null;
+        public static void main(String args[]) {
```
looks like this is indented too much - I think it should be 4 spaces instead of 8

serpient committed:
Update folder structure and spacing

- Move Mock classes into the testing folder as it is only used in Tests
- Fix spacing in a few files.

ssunday commented on src/main/java/java_echo_server/App.java:
```diff
@@ -4,19 +4,19 @@
-            } catch (IOException e) {
-                System.err.println(e.toString());
-            }
+    public static void main(String args[]) {
```
yay spacing

serpient committed:
Update ClientTest to test ClientSocket

The ClientTest is currently using MockClientSocket to meet the test assertions. While MockClientSocket is good for the integration tests in ServerTest, it is not a replacement for actually testing ClientSocket. This commit updates the codebase to allow for testing of ClientSocket which requires dependency injection of the original client, input and output streams.

- Update ClientSocket to take in a clientSocket, input and output streams during initialization. This allows us to pass in a mock input and output during testing.
- Update EchoProtocol to create input and output streams in the run() and pass into ClientSocket as parameters
- Update ClientTest to use ClientSocket instead of MockClientSocket to pass the assertions.
